packages/expo-cli/src/commands/__tests__/upgrade-test.ts

```diff
@@ -0,0 +1,319 @@
+      '/from-app-json-UNVERSIONED/app.json': JSON.stringify({
+        expo: { ...appJson, sdkVersion: 'UNVERSIONED' },
+      }),
+      '/from-app-json-UNVERSIONED/app.config.js': `module.exports=({config}) => ({ ...config, sdkVersion: '34.0.0' })`,
+      '/from-app-json-UNVERSIONED/node_modules/expo/package.json': expoPackageJson,
+      '/from-app-json-UNVERSIONED/node_modules/expo/bundledNativeModules.json': JSON.stringify({}),
+
+      '/from-app-json/package.json': packageJson,
+      '/from-app-json/app.json': JSON.stringify({ expo: { ...appJson } }),
+      '/from-app-json/app.config.js': `module.exports=({config}) => ({ ...config, sdkVersion: '34.0.0' })`,
+      '/from-app-json/node_modules/expo/package.json': expoPackageJson,
+      '/from-app-json/node_modules/expo/bundledNativeModules.json': JSON.stringify({}),
+
+      '/from-app-config-js/package.json': packageJson,
+      '/from-app-config-js/app.json': JSON.stringify({
+        expo: { ...appJson, sdkVersion: undefined },
+      }),
+      '/from-app-config-js/app.config.js': `module.exports=({config}) => ({ ...config, sdkVersion: '34.0.0' })`,
+      '/from-app-config-js/node_modules/expo/package.json': expoPackageJson,
+      '/from-app-config-js/node_modules/expo/bundledNativeModules.json': JSON.stringify({}),
+
+      '/from-expo-package/package.json': packageJson,
+      '/from-expo-package/app.json': JSON.stringify({
+        expo: { ...appJson, sdkVersion: undefined },
+      }),
+      '/from-expo-package/app.config.js': `module.exports=({config}) => ({ ...config })`,
+      '/from-expo-package/node_modules/expo/package.json': expoPackageJson,
+      '/from-expo-package/node_modules/expo/bundledNativeModules.json': JSON.stringify({}),
+    });
+  });
+  afterAll(() => {
+    vol.reset();
+  });
+
+  // Ensure we delete the app.json sdkVersion if it's defined
+  it(`delete the app.json sdkVersion if it's defined`, async () => {
+    require('@expo/config').writeConfigJsonAsync = jest.fn((...props) => {
+      return jest.requireActual('@expo/config').writeConfigJsonAsync(...props);
+    });
+    const projectRoot = '/from-app-json';
+    await upgradeAsync(
+      { projectRoot, workflow: 'bare', requestedSdkVersion: null },
+      { yarn: true }
+    );
+    expect(require('@expo/config').writeConfigJsonAsync).toBeCalledTimes(1);
+    expect(require('@expo/config').writeConfigJsonAsync).toHaveBeenLastCalledWith(projectRoot, {
+      sdkVersion: undefined,
+    });
+
+    const { exp, rootConfig: json } = await require('@expo/config').readConfigJsonAsync(
+      projectRoot
+    );
+    expect(json.expo.sdkVersion).not.toBeDefined();
+    // Uses expo package version
+    expect(exp.sdkVersion).toBe('35.0.0');
+  }, 5000);
+
+  // Important to ensure that we don't modify the app.json extraneously
+  it(`skips modifying the app.json if the app.json doesn't define an sdkVersion`, async () => {
+    require('@expo/config').writeConfigJsonAsync = jest.fn();
```

Comment:
this will change in https://github.com/expo/expo-cli/pull/1584 because we may alter other things about `app.json`

Reply:
We should limit the amount of writing we do to the app.json. We can't/shouldn't write to the app.config.js and having both an app.json and app.config.js makes using a custom config path difficult (if not impossible).

Reply:
I also think we should merge this before #1584 and update the tests there to reflect breaking changes in the expected behavior.

Reply:
i agree we should limit changes. we should also perhaps write #1584 such that we warn users if a value is defined in app.config.js - 1) use readConfig to get the final exp value, 2) check if key exists in app.json, then 3) check if it exists in final exp value 4) if key does not exist in step 2 but does exist in step 3 then we warn the user that they should update it.. or something like that